Classes/Command/NodeIndexCommandController.php

```diff
@@ -234,131 +196,377 @@ public function indexNodeCommand(string $identifier, string $workspace = null):
-                $this->outputLine('<error>Error</error> ' . $error->message());
+    /**
+     * @param string $workspace
+     * @param string $dimensionsValues
+     * @param int $postfix
+     * @param int $limit
+     * @return int
+     * @Flow\Internal
+     */
+    public function buildWorkspaceInternalCommand($workspace, $dimensionsValues, $postfix, $limit = null)
+    {
+        $dimensionsValues = $this->configureInternalCommand($dimensionsValues, $postfix);
+
+        $workspaceLogger = function ($workspaceName, $indexedNodes, $dimensions) {
+            if ($dimensions === []) {
+                $message = 'Workspace "' . $workspaceName . '" without dimensions done. (Indexed ' . $indexedNodes . ' nodes)';
+            } else {
+                $message = 'Workspace "' . $workspaceName . '" and dimensions "' . json_encode($dimensions) . '" done. (Indexed ' . $indexedNodes . ' nodes)';
             }
-            $this->outputLine();
-            $this->outputLine('<error>Check your logs for more information</error>');
-        } else {
-            $this->logger->log('Done. (indexed ' . $count . ' nodes)', LOG_INFO);
-        }
+            $this->outputLine($message);
+        };
+
+        $count = $this->worksaceIndexer->indexWithDimensions($workspace, $dimensionsValues, $limit, $workspaceLogger);
+
+        $this->outputMemoryUsage();
+        $this->outputErrorHandling();
+        $this->outputLine('Number of nodes: %s', [$count]);
+    }
+
+    /**
+     * @param string $dimensionsValues
+     * @param int $postfix
+     * @Flow\Internal
+     */
+    public function refreshInternalCommand($dimensionsValues, $postfix)
+    {
+        $this->configureInternalCommand($dimensionsValues, $postfix);
+
+        $this->logger->log(vsprintf('Refresh index %s', [$this->nodeIndexer->getIndexName()]), LOG_INFO);
+        $this->outputMemoryUsage();
+
         $this->nodeIndexer->getIndex()->refresh();
 
-        // TODO: smoke tests
-        if ($update === false) {
-            $this->nodeIndexer->updateIndexAlias();
+        $this->outputErrorHandling();
+    }
+
+    /**
+     * @param string $dimensionsValues
+     * @param int $postfix
+     * @param bool $update
+     * @Flow\Internal
+     */
+    public function aliasInternalCommand($dimensionsValues, $postfix, $update = false)
+    {
+        if ($update === true) {
+            return;
+        }
+        $this->configureInternalCommand($dimensionsValues, $postfix);
+
+        $this->logger->log(vsprintf('Update alias for index %s', [$this->nodeIndexer->getIndexName()]), LOG_INFO);
+        $this->nodeIndexer->updateIndexAlias();
+        $this->outputErrorHandling();
+    }
+
+    /**
+     * @param string|array $dimensionsValues
+     * @param string $postfix
+     * @return array
+     */
+    protected function configureInternalCommand($dimensionsValues, $postfix)
+    {
+        if (!\is_array($dimensionsValues)) {
```

Comment:
Seems this method is called with a json string only. So this if can be removed and the method signature should be defined more strict.

Reply:
No in the build method it's called with an array